lib/Sema/CSDiagnostics.h

```diff
@@ -91,6 +91,13 @@ class FailureDiagnostic {
   Type getType(Expr *expr, bool wantRValue = true) const;
   Type getType(const TypeLoc &loc, bool wantRValue = true) const;
 
+  /// For a given locator describing a function argument conversion, or a
+  /// constraint within an argument conversion, returns information about the
+  /// application of the argument to its parameter. If the locator is not
+  /// for an argument conversion, returns \c None.
+  static Optional<FunctionArgApplyInfo>
+  getFunctionArgApplyInfo(ConstraintSystem &cs, ConstraintLocator *locator);
```

Comment:
I don't think removing the dependency on `ConstraintSystem` and using `Solution` instead necessarily means that we'd need to make it an instance method again. The methods that we call on `ConstraintSystem` in `getFunctionArgApplyInfo` are either already are implemented or can be implemented on `Solution`. I don't see much of a difference between passing in a `ConstraintSystem` vs calling `getConstraintSystem()` on the first line - we'll have to solve the same problems either way when we make the switch over to `Solution` because we call methods directly on `ConstraintSystem` here, even before my changes. I think it would even make sense to implement this on `Solution`, since it's really just computing information about a particular solution and isn't related to `FailureDiagnostic` otherwise.

That said, the first thing I tried was implementing it inline. It needs so much of the logic from this function - everything up until the point where we find the interface type of the function because the first ~60 lines of code are just for finding the function type

Reply:
If we want to keep using it that way it would have to be moved to constraint system itself then because it no longer belongs to the diagnostic.

Reply:
This also points to the bigger problem with `AllowTupleTypeMismatch` fix - fixes supposed to be recording useful information and in cases where locator is pointing to argument/parameter comparison `repairFailures` would already provide  both argument and parameter types...

Reply:
The problem is that when we're in `repairFailures` due to a tuple element mismatch for an argument to a function, we don't have the full argument and parameter type. We only fail at the particular element, so we don't actually have those types when we create the fix.

Reply:
Alright, let's just move this logic to be a member of constraint system until we figure out a better solution.